app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -70,175 +41,107 @@ private const val MOAT_FRONT = "ajax.aspnetcdn.com"
      */
-    @Suppress("BlockingMethodInNonBlockingContext")
     suspend fun start() = withContext(Dispatchers.IO) {
         if (state.value !is TorState.Stopped) stop()
 
         LOG.info("Starting...")
         val now = System.currentTimeMillis()
         _state.value = TorState.Starting(progress = 0, startTime = now, lastProgressTime = now)
-        startLatch = CountDownLatch(1)
-        app.registerReceiver(broadcastReceiver, IntentFilter(ACTION_STATUS))
-        app.registerReceiver(errorReceiver, IntentFilter(ACTION_ERROR))
-        broadcastReceiverRegistered = true
 
         Intent(app, ShareService::class.java).also { intent ->
             startForegroundService(app, intent)
         }
-        try {
-            executableManager.installObfs4Executable()
-            executableManager.installSnowflakeExecutable()
-            startLatch?.await() ?: error("startLatch was null")
-            startLatch = null
-            onTorServiceStarted()
-        } catch (e: Exception) {
-            LOG.warn("Error while starting Tor: ", e)
-            stop()
-        }
+
+        val torWrapper = TorModule.createTorWrapper(app, wakeLockManager)
+        tor = torWrapper
+        torWrapper.setObserver(this@TorManager)
+        torWrapper.start()
+        torWrapper.enableNetwork(true)
     }
 
     fun stop(failedToConnect: Boolean = false) {
         LOG.info("Stopping...")
         startCheckJob?.cancel()
         startCheckJob = null
         _state.value = TorState.Stopping(failedToConnect)
-        controlConnection = null
+        tor?.stop()
+        tor = null
         Intent(app, ShareService::class.java).also { intent ->
             app.stopService(intent)
         }
-        // wait for service to stop and broadcast when it gets destroyed
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    override fun onState(s: TorWrapper.TorState) {
+        when (s) {
+            STARTING_STOPPING -> onStopped()
+            CONNECTING -> GlobalScope.launch {
+                LOG.info("connecting...")
+                try {
+                    onTorServiceStarted()
+                } catch (e: Exception) {
+                    LOG.warn("Error while starting Tor: ", e)
+                    stop()
+                }
+            }
+
+            CONNECTED -> LOG.info("connected.")
+            DISABLED -> LOG.info("disabled.")
+        }
     }
 
     @UiThread
     fun onStopped() {
-        if (broadcastReceiverRegistered) {
-            app.unregisterReceiver(broadcastReceiver)
-            app.unregisterReceiver(errorReceiver)
-            broadcastReceiverRegistered = false
-        }
-        localSocket = null
         LOG.info("Stopped")
         val failedToConnect = (state.value as? TorState.Stopping)?.failedToConnect == true
         _state.value = TorState.Stopped(failedToConnect)
     }
 
+    override fun onBootstrapPercentage(percentage: Int) {
+        changeStartingState(percentage)
```

Comment:
Bootstrapping 100% used to translate to 70% progress, but now it looks like it translates to 100%.

Reply:
Oops, good catch!